### DISCUSSION TITLE
Is it possible to run Bear via manual LD_PRELOAD?

### DISCUSSION THREAD

**Discussion by yugr** [rizsotto/Bear#442] [Q&A] [answered]
I'm trying to collect compilation database in a package build system (`pbuilder`) environment. Unfortunately the build infrastructure does not provide control over build command so I can not simply apply `bear` to it:
```
$ bear make clean all
```
I do have control over environment variables though so I could set `LD_PRELOAD` by hand prior to running the build. I this possible? If yes, could this scenario be added to docs?

**Comment by rizsotto** [comment, marked as answer]
Hey @yugr , this is a good question. I don't think the `LD_PRELOAD` itself can help you out to solve the `pbuilder` limitation. (Definitely not with the 3.x versions, maybe with the 2.4 version.)

- one of the workaround can be to run the `pbuilder` command itself inside Bear. `bear --config config.json -- pbuilder ...` The config file shall say that `output.content.include_only_existing_source` should set to false, and the `output.content.paths_to_include` should be set to the source directory of your project. (I would run first without the `paths_to_include` set, to see which files it lists.) You can speed up this process by running the `intercept` command once against the build, and just run subsequent `citnames` to customize the output.
- the other workaround I can imagine is to create a fake tool directory. Just create a directory which contains a new `make` program, which calls `bear -- /usr/bin/make $@`, and set your `PATH` to have this directory in the front. Then run `pbuilder` which is likely to pick up fake `make` program.

I made these suggestion without really know how `pbuilder` works. (I did try this with arch linux package maker though.)

**Comment by yugr** [comment]
Thank you for the suggestion, for `make` it would indeed work. But it won't help with other build systems: `ninja`, `scons`, `python setuptools`, `Perl-Build` or even custom shell scripts (Debian has over 50K OSS packages and you'd be  surprised how inventive people are). In theory all of those would need to be instrumented which does not make it a scalable solution.
